jxl/src/render/stages/noise.rs

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// TODO(zond): remove once we use this!
+#![allow(dead_code, unused_variables)]
+
+use crate::{
+    features::noise::Noise,
+    render::{RenderPipelineInOutStage, RenderPipelineInPlaceStage, RenderPipelineStage},
+};
+
+pub struct ConvolveNoiseStage {
+    first_channel: usize,
+}
+
+impl ConvolveNoiseStage {
+    pub fn new(first_channel: usize) -> ConvolveNoiseStage {
+        ConvolveNoiseStage { first_channel }
+    }
+}
+
+impl std::fmt::Display for ConvolveNoiseStage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "convolve noise for channels [{},{},{}]",
+            self.first_channel,
+            self.first_channel + 1,
+            self.first_channel + 2
+        )
+    }
+}
+
+impl RenderPipelineStage for ConvolveNoiseStage {
+    type Type = RenderPipelineInOutStage<f32, f32, 2, 2, 0, 0>;
+
+    fn uses_channel(&self, c: usize) -> bool {
+        c >= self.first_channel && c < self.first_channel + 3
+    }
+
+    fn process_row_chunk(
+        &mut self,
+        position: (usize, usize),
+        xsize: usize,
+        row: &mut [(&[&[f32]], &mut [&mut [f32]])],
+    ) {
+        for (input, output) in row {
+            for x in 0..xsize {
+                let mut others = 0.0;
+                for i in 0..5 {
+                    let offset = (x as i32 + i) as usize;
+                    others += input[0][offset];
+                    others += input[1][offset];
+                    others += input[3][offset];
+                    others += input[4][offset];
+                }
+                others += input[2][x];
+                others += input[2][x + 1];
+                others += input[2][x + 3];
+                others += input[2][x + 4];
+                output[0][x] = others * 0.16 + input[2][x + 2] * -3.84;
+            }
+        }
+    }
+}
+
+pub struct AddNoiseStage {
+    noise: Noise,
```

Comment:
Any reason not to have this be a StrengthEvalLut which owns Noise?

Reply:
I moved the entire strength calculation to the Noise struct instead, that makes more sense IMO.